In [2]:
import numpy as np
import pandas as pd

In [3]:
s=pd.read_csv("train.csv")

In [4]:
s.head(10)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   
5            6         0       3   
6            7         0       1   
7            8         0       3   
8            9         1       3   
9           10         1       2   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   
5                                   Moran, Mr. James    male  45.0      0   
6                            McCarthy, Mr. Timothy J    male  54.0      0   
7                     Palsson, Master. Gosta Leonard    male   2.0      3   
8  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S  
5      0            330877   8.4583   NaN        Q  
6      0             17463  51.8625   E46        S  
7      1            349909  21.0750   NaN        S  
8      2            347742  11.1333   NaN        S  
9      0            237736  30.0708   NaN        C

In [5]:
s.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age              0
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         0
dtype: int64

In [6]:
from sklearn import preprocessing
l_encoder=preprocessing.LabelEncoder()
encoded_g=l_encoder.fit_transform(s['Sex'])
s['Sex']=encoded_g
s['Embarked']=l_encoder.fit_transform(s['Embarked'])

In [8]:
s=s.drop(['Cabin','PassengerId','Name','Ticket'],axis=1)
s.head()

Survived  Pclass  Sex   Age  SibSp  Parch     Fare  Embarked
0         0       3    1  22.0      1      0   7.2500         2
1         1       1    0  38.0      1      0  71.2833         0
2         1       3    0  26.0      0      0   7.9250         2
3         1       1    0  35.0      1      0  53.1000         2
4         0       3    1  35.0      0      0   8.0500         2

## implementing naive bayes

In [9]:
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import BernoulliNB
clf=BernoulliNB()
from sklearn.metrics import accuracy_score,confusion_matrix

In [10]:
def naive_b(a,b):
    X_train,X_test,y_train,y_test=train_test_split(a,b,test_size=0.3,random_state=0)
    y_pred=clf.fit(X_train,y_train).predict(X_test)
    score=accuracy_score(y_test,y_pred,normalize=True)
    cm=confusion_matrix(y_test,y_pred)
    return score,cm

#### Pclass vs remaining attributes

In [14]:
y=s.iloc[:,1:2]
X=s.drop(["Pclass"],axis=1)

In [16]:
score1,cm1=naive_b(X,y)
print("The accuracy score is:",score1)
print("The counfusion matrix is:")
cm1

The accuracy score is: 0.5917602996254682
The counfusion matrix is:


C:\ProgramData\Anaconda3\lib\site-packages\sklearn\utils\validation.py:760: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


array([[ 30,   8,  32],
       [  9,  10,  30],
       [ 24,   6, 118]], dtype=int64)

#### 158 records out of total 267 records are successfully classified, the model accuracy is 59.17%

### attribute sex vs the remaining attributes

In [23]:
y=s.iloc[:,2:3]
X=s.drop(["Sex"],axis=1)

In [24]:
score2,cm2=naive_b(X,y)
print("The accuracy score is:",score2)
print("The counfusion matrix is:")
print(cm2)

The accuracy score is: 0.7453183520599251
The counfusion matrix is:
[[ 49  49]
 [ 19 150]]


C:\ProgramData\Anaconda3\lib\site-packages\sklearn\utils\validation.py:760: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


#### 199 records out of total 267 records are successfully classified, the model accuracy is 74.5%


### attribute SibSp vs the remaining attributes

In [27]:
y2=s.iloc[:,4:5]
X2=s.drop(["SibSp"],axis=1)

In [29]:
score3,cm3=naive_b(X2,y2)
print("The accuracy score is:",score3)
print("The counfusion matrix is:")
print(cm3)

The accuracy score is: 0.6891385767790262
The counfusion matrix is:
[[162  20   0   0   0   0   0]
 [ 43  22   0   0   0   0   0]
 [  6   2   0   0   0   0   0]
 [  5   2   0   0   0   0   0]
 [  2   0   0   0   0   0   0]
 [  1   0   0   0   0   0   0]
 [  2   0   0   0   0   0   0]]


C:\ProgramData\Anaconda3\lib\site-packages\sklearn\utils\validation.py:760: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


#### 184 records out of total 267 records are successfully classified, the model accuracy is 68.9%

### parch vs other remaining attribute

In [31]:
y3=s.iloc[:,5:6]
X3=s.drop(["Parch"],axis=1)

In [33]:
score4,cm4=naive_b(X3,y3)
print("The accuracy score is:",score4)
print("The counfusion matrix is:")
print(cm4)

The accuracy score is: 0.7153558052434457
The counfusion matrix is:
[[182  16   0   0   0]
 [ 31   9   0   0   0]
 [ 24   3   0   0   0]
 [  1   0   0   0   0]
 [  1   0   0   0   0]]


C:\ProgramData\Anaconda3\lib\site-packages\sklearn\utils\validation.py:760: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


#### 191 records out of total 267 records are successfully classified, the model accuracy is 71.5%

## Embarked vs all the remaining attributes

In [37]:
y4=s.iloc[:,7:]
X4=s.drop(["Embarked"],axis=1)

In [38]:
score5,cm5=naive_b(X4,y4)
print("The accuracy score is:",score5)
print("The counfusion matrix is:")
print(cm5)

The accuracy score is: 0.7340823970037453
The counfusion matrix is:
[[  0   0  49]
 [  0   0  22]
 [  0   0 196]]


C:\ProgramData\Anaconda3\lib\site-packages\sklearn\utils\validation.py:760: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


#### 196 records out of total 267 records are successfully classified, the model accuracy is 73.4%